### PR TITLE
Properly handling custom op exception by modify engine

### PR DESCRIPTION
**Pull request by arcadiaphy**
## Description ##
In #14522, problems of custom op exception handling are reported. #14575 tries to fix them by modify custom op implementation, which causes blocking in custom thread.

This PR tries to address the issue by modifying engine:

1. Add kNoSkip operation to force execution even with associated exception
2. Since NDArray is duplicated in custom op with shared data, their exception status may not be synchronized,  so `Throw` method is added in engine to re-throw associated exception on var.
3. Users are free to use sync operation like `asnumpy` or `wait_to_read` in custom op to block custom thread, so unlimited number of threads are still needed to avoid deadlock.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szha** [comment]
@junrushao1994 could you take a look too?

File: src/engine/threaded_engine.cc [hunk 1]
```diff
@@ -498,6 +498,11 @@ inline void ThreadedEngine::ThrowException(ThreadedVar* threaded_var) {
   return;
 }
 
+void ThreadedEngine::Throw(VarHandle var) {
+  ThreadedVar *threaded_var = ThreadedVar::CastFromBase(var);
+  ThrowException(threaded_var);
+}
```
Comment (anirudh2290) [inline src/engine/threaded_engine.cc]: We dont need this in engine, please see above.

File: src/operator/custom/custom-inl.h [hunk 2]
```diff
@@ -116,6 +121,16 @@ class CustomOperator {
 
       Engine::Get()->PushSync(
           [=](RunContext rctx) {
+            try {
+              Throw();
+              for (const auto& i : arrs) {
+                Engine::Get()->Throw(i.var());
+              }
```
[the remark this answers is not in the document]
Comment (arcadiaphy) [reply src/operator/custom/custom-inl.h]: Actually, we need this. In custom op, outside ndarray are duplicated with shared data by `AllocateNDArrayCopy`, so their exception status may not be synchronized.

A possible scenario is: custom op is successfully pushed to engine but get OOM error in real execution. Then inside ndarray will have exception set, meanwhile its outside duplicated array has no exception set. When you call sync function on outside ndarray, the exception will be silently ignored.

Two solution here:
1. re-throw in CustomOperator
2. synchronize exception status by copy inside ndarray's exception to outside ndarray

I use the 1st solution here.

[the remark this answers is not in the document]
Comment (arcadiaphy) [reply src/operator/custom/custom-inl.h]: Let me post a example here:
```
import mxnet as mx


class MyMulMax(mx.operator.CustomOp):
    def __init__(self):
        super(MyMulMax, self).__init__()

    def forward(self, is_train, req, in_data, out_data, aux):
        a, b = in_data[0:2]
        c = mx.nd.batch_dot(a, b)
        d = mx.nd.max(c, axis=-1, keepdims=True)
        self.assign(out_data[0], req[0], d)

    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        self.assign(in_grad[0], req[0], 0)
        self.assign(in_grad[1], req[1], 0)


@mx.operator.register("MyMulMax")
class MyMulMaxProp(mx.operator.CustomOpProp):
    def __init__(self):
        super(MyMulMaxProp, self).__init__()

    def list_arguments(self):
        return ['a', 'b']

    def list_outputs(self):
        return ['d']

    def infer_shape(self, in_shape):
        return in_shape, [list(in_shape[0][:-1] + [1])]

    def create_operator(self, ctx, shapes, dtypes):
        return MyMulMax()

def custom(n):
    with mx.Context(mx.gpu()):
        a = mx.nd.random.uniform(shape=(n, 600, 1))
        b = mx.nd.random.uniform(shape=(n, 1, 700))
        c = mx.nd.Custom(a, b, op_type="MyMulMax")
        # here c is outside ndarray, its duplicated ndarray is
        # out_data[0] in custom op with shared data,
        # their associated exception may not be synchronized
        # when OOM happens inside custom op
        c.wait_to_read()


if __name__ == "__main__":
    custom(100)
```

[the remark this answers is not in the document]
Comment (anirudh2290) [reply src/operator/custom/custom-inl.h]: makes sense. thanks for the example!

File: tests/python/unittest/test_operator.py [hunk 3]
```diff
@@ -5355,29 +5356,29 @@ def create_operator(self, ctx, shapes, dtypes):
...
-                return ['output']
-            def infer_shape(self, in_shape):
-                return in_shape, [in_shape[0]]
-            def create_operator(self, ctx, shapes, dtypes):
-                return AdditionOP()
+    class AdditionOP(mx.operator.CustomOp):
+        def __init__(self):
+            super(AdditionOP, self).__init__()
+        def forward(self, is_train, req, in_data, out_data, aux):
+            out_data[0][:] = in_data[0] + in_data[1]
```
Comment (anirudh2290) [inline tests/python/unittest/test_operator.py]: can we also add a test for a forward which has wait_to_read on out_data.

Comment (arcadiaphy) [reply tests/python/unittest/test_operator.py]: ok, I'll add more tests to cover all scenarios

**Comment by arcadiaphy** [comment]
irrelevant flaky test in julia:
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-cpu/detail/PR-14693/4/pipeline

**Comment by TaoLv** [comment]
@arcadiaphy Can you take a look at this failure? http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-cpu/detail/PR-14783/3/pipeline#step-676-log-1586

**Comment by arcadiaphy** [comment]
@TaoLv Sure, let me have a look.

**Pull request merged**